Guide/src/reference/architecture/openhcl_servicing.md

```diff
@@ -0,0 +1,173 @@
+
+**Without Keepalive:**
+- Devices are cleanly shut down
+- Device handles are closed
+- DMA allocations are released
+- VFIO device handles are dropped (causing device reset via FLR)
+
+**With Keepalive (NVMe):**
+- NVMe devices remain connected (CC.EN=1)
+- VFIO device handles are kept open (preventing Function Level Reset)
+- DMA buffers in the private pool are preserved
+- Device maintains its operational state
+
+### 4. Host Servicing
+
+The host can prepare for servicing while OpenHCL is still running by loading the new IGVM file into memory and preparing updated binaries. Once OpenHCL is stopped, the host applies the updates.
+
+During the blackout period:
+- VM guest state in VTL0 is preserved by the host
+- OpenHCL binaries and kernel can be updated
+- The new IGVM file is applied
+
+### 5. OpenHCL Restart
+
+The host restarts OpenHCL with a new boot flow (see [OpenHCL Boot Process](./openhcl_boot.md)) using the new version and provides the saved state as input.
+
+### 6. State Restoration
+
+OpenHCL restores components in dependency order. The DMA Manager is restored first to provide memory pools for other components, followed by device-specific managers and state units. For complete details on the restoration process and ordering, see the implementation in `underhill_core`.
+
+For details on NVMe keepalive behavior during servicing, see [NVMe Storage Backend](../../backends/storage/nvme.md).
+
+## Compatibility and Versioning
+
+### Saved State Format
+
+The servicing state uses Protocol Buffers for serialization. For details on save state implementation, compatibility requirements, and best practices, see the [Save State guide](../../dev_guide/contrib/save-state.md).
+
+## Configuration
+
+Servicing behavior is controlled by several environment variables and configuration parameters. These parameters are set by the boot shim and passed to the kernel, init, and OpenHCL processes.
+
+### NVMe Keepalive Configuration
+
+The `OPENHCL_NVME_KEEP_ALIVE` environment variable controls NVMe keepalive behavior. For details on the values and their meanings, see the [KeepAliveConfig rustdocs](https://openvmm.dev/rustdoc/linux/underhill_core/options/enum.KeepAliveConfig.html).
+
+The boot shim (see `openhcl_boot`) infers the configuration based on the detected environment unless explicitly overridden.
+
+### DMA Pool Configuration
+
+The `OPENHCL_IGVM_VTL2_GPA_POOL_CONFIG` parameter controls the VTL2 GPA pool configuration used for the private pool. For details on the parameter values and behavior, see the [Vtl2GpaPoolConfig rustdocs](https://openvmm.dev/rustdoc/linux/openhcl_boot/struct.Vtl2GpaPoolConfig.html).
+
+The boot shim determines pool sizes using heuristics defined in `openhcl_boot` based on the system configuration, unless explicitly overridden.
+
+### Test Scenarios
+
+For testing servicing behavior, the `OPENHCL_TEST_CONFIG` environment variable can be set to simulate specific conditions:
```

Comment:
Here, and elsewhere you talk about setting env config, link to instructions on setting the OpenHCL command line if found. If the Guide does not address that, clarify that these cmdline parameters MUST be sent to the paravisor.